### PR TITLE
feat(wasm,geometry): expose general 2D Boolean ops over contour sets (#1863)

### LINKED ISSUE #1863: Expose general 2D Boolean ops (union / difference / intersection over contour sets) from `@ifc-lite/wasm`
### The problem

I'm implementing an analytic hidden-surface-removal renderer ("Design 2", a front-to-back cumulative occluder) for the ortho SVG viewer, following your guidance in [discussion #1122](vscode-file://vscode-app/c:/Users/A524573/AppData/Local/Programs/Microsoft%20VS%20Code/8a7abeba6e/resources/app/out/vs/code/electron-browser/workbench/workbench.html). The core loop needs general 2D Boolean operations on contour sets:
```rust
vis = outline(e) - C        # difference
C   = C ∪ outline(e)        # union
clip = outline(e) ∩ tile    # intersection (screen tiling)
```
Today the only `i_overlay` capability that crosses the WASM boundary is the fixed-purpose `meshOutline2d`. The general primitives already exist in [rust/geometry/src/bool2d.rs](https://github.com/LTplus-AG/ifc-lite/blob/main/rust/geometry/src/bool2d.rs) (`subtract_2d`, `subtract_multiple_2d`, `union_contours_to_shapes`, `i_overlay-backed`) but none are `#[wasm_bindgen]`, so they're unreachable from `@ifc-lite/wasm`. This forces me onto the JS `polygon-clipping` package, which means a second geometry engine with different robustness/winding/precision semantics than `meshOutline2d`, plus JS-side coordinate-scaling workarounds. I can't round-trip `meshOutline2d` output through a Boolean op without leaving the library.

### Proposed solution

Expose the existing bool2d.rs primitives across the WASM boundary, as offered at [discussion #1122](vscode-file://vscode-app/c:/Users/A524573/AppData/Local/Programs/Microsoft%20VS%20Code/8a7abeba6e/resources/app/out/vs/code/electron-browser/workbench/workbench.html)

A minimal contour-set-oriented surface (names/shapes open to your conventions):

- `union2d(a, b) / unionContoursToShapes(contours)`
- `difference2d(a, b)`, `wraps subtract_2d`
- `differenceMultiple2d(a, subtrahends)`, wraps `subtract_multiple_2d`
- `intersection2d(a, b)`

Acceptance criteria:

- [ ]  The three existing primitives (or equivalents), plus intersection, callable from `@ifc-lite/wasm`.
- [ ]  I/O uses a documented contour-set type interoperable with `meshOutline2d` output, so its result can be fed directly into a Boolean op.
- [ ]  Multi-polygons with holes and disjoint islands supported; winding stays compatible with fill-rule="nonzero" (consistent with `meshOutline2d`).
- [ ]  Degenerate inputs (empty operands, coincident/near-collinear edges) return empty results rather than panicking.
- [ ]  TS typings shipped with the package.
- [ ]  Target release noted so I can pin the submodule (currently at [68b2703](https://github.com/LTplus-AG/ifc-lite/commit/68b27030f41659c8046e1eb1bfada8ad32cef52c) / @ifc-lite/geometry@2.7.9).

### Alternatives considered

- JS `polygon-clipping` package (current fallback). Works, but introduces a second geometry engine alongside i_overlay with different precision/winding behavior, requires manual coordinate scaling for robustness, and prevents clean round-tripping of `meshOutline2d` output. I'm shipping Design 2 behind an abstraction layer with this as the default backend so I can switch to the WASM backend via config once the bindings land, so this request is non-blocking short-term.
- Reusing `meshOutline2d` alone. Insufficient: it only unions the triangles of a single element into a silhouette; it can't difference one element's silhouette against an accumulated occluder or intersect against a tile rectangle.
- Reimplementing the Booleans in JS/Rust myself. Wasteful and divergent. The robust, i_overlay-backed implementation already exists in `bool2d.rs`; the only gap is the `#[wasm_bindgen]` surface.

**Note**: [subtract_2d](https://github.com/LTplus-AG/ifc-lite/blob/68b27030f41659c8046e1eb1bfada8ad32cef52c/rust/geometry/src/bool2d.rs#L38) and [subtract_multiple_2d](https://github.com/LTplus-AG/ifc-lite/blob/68b27030f41659c8046e1eb1bfada8ad32cef52c/rust/geometry/src/bool2d.rs#L77) are confirmed in bool2d.rs in the version my submodule is pinned to (@ifc-lite/geometry@2.7.9); `union_contours_to_shapes` I'm taking from your comment, as it isn't in that pinned version yet.

### PR DESCRIPTION
**Pull request by louistrue**
Closes #1863.

## What

Exposes general 2D boolean operations over **contour sets** from `@ifc-lite/wasm`: `union2d`, `difference2d`, `intersection2d`, `resolve2d`, and the `Contours2D` handle they operate on. `Contours2D.fromMeshOutline(outline)` adopts a `meshOutline2d` result directly.

Until now the only `i_overlay` capability crossing the wasm boundary was the fixed-purpose `meshOutline2d` (unions one mesh's projected triangles into a silhouette). @mehmet-ylcnky's analytic HSR renderer ("Design 2", a front-to-back cumulative occluder) needs to combine two silhouettes — `vis = outline(e) - C`, `C = C ∪ outline(e)`, `clip = outline(e) ∩ tile` — which forced a second JS geometry engine (`polygon-clipping`) with different winding/precision semantics alongside the outlines it consumes.

## The one design decision worth flagging

The internal `bool2d::subtract_2d` / `subtract_multiple_2d` are **not** wrapped directly. They are `Profile2D`-shaped and collapse their result to the largest output shape (`shapes_to_profile`). That is right for the single extrusion profile they serve and **silent geometry loss** for HSR: a wall seen behind a column is two visible slivers, and `outline(e) - C` legitimately splits into many disjoint islands. So this adds a general layer (`contour_bool2d.rs`) that keeps **every** disjoint output shape with its holes, grouped via `shapeOffsets()`.

## Contract (per the acceptance criteria in #1863)

- [x] Union / difference / intersection over contour sets, callable from `@ifc-lite/wasm`.
- [x] I/O uses `Contours2D`, interoperable with `meshOutline2d` output (`fromMeshOutline`), so a result feeds straight into another boolean.
- [x] Multi-polygons with holes and disjoint islands preserved; winding stays `fill-rule="nonzero"`-compatible (CCW covers, CW subtracts, input winding **respected not normalised** — normalising is exactly what destroys holes).
- [x] Degenerate input (empty operands, < 3 vertices, non-finite coords, explicit closing vertex) returns empty rather than panicking; every empty-operand combination has a defined answer.
- [x] TS typings shipped (committed `pkg/ifc-lite.d.ts`).
- [x] **Target release**: `@ifc-lite/wasm` **minor** (changeset included). Pin the submodule to the release that ships from this once merged.

`difference2d` takes any number of clip rings (they subtract as their union under NonZero), so there is no separate `differenceMultiple2d` — one entry point covers it.

## Memory ownership

Operations return **new** `Contours2D` handles and never mutate their operands. An accumulating occluder loop must `free()` the handle it replaces; the JS doc comment on the module shows the pattern. `bounds()` is provided as a cheap pre-test (skip the boolean when an accumulated occluder's bounds miss the next element).

## Tests

- 21 Rust unit tests (`contour_bool2d_tests.rs`): union/difference/intersection, island preservation, hole punching + winding, accumulator == single-union, round-trip stability, `meshOutline2d` interop, degenerate input, accessors.
- 8 real-boundary cases in `test-wasm-contract.mjs` — area is read back **across** the wasm boundary as the sum of signed ring areas, which is the covered total only if winding still carries outer-vs-hole, so a boundary that ever normalised winding would fail these.

Rust: `cargo test -p ifc-lite-geometry --lib` 532 pass. Boundary: `test-wasm-contract.mjs` 35 pass. Gates: api-surface snapshot updated (+5 exports), committed wasm d.ts regenerated, wasm-free typecheck 33/33, module-size ratchet green (both new files < 400 lines).

## Perf verdict

No hot-path change. `contour_bool2d` is a new leaf module reached only through the new wasm exports; nothing in the parse/geometry batch pipeline calls it. The operations are `i_overlay` (the same engine the void/CSG paths already use), f64, deterministic (native == wasm byte-identical).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added `Contours2D` for building/inspecting 2D contour ring sets and managing WASM memory (`free`/dispose).
  * Added general 2D contour boolean operations: `union2d`, `difference2d`, `intersection2d`, and `resolve2d`.
  * Added conversion from `meshOutline2d` to `Contours2D` for boolean workflows.
* **Bug Fixes**
  * Improved handling of empty and degenerate inputs with well-defined, non-crashing results.
* **Tests**
  * Added unit and WASM contract coverage for winding/holes, disjoint islands, and empty cases.
* **Documentation**
  * Updated WASM API docs with the new `Contours2D` section and operations.

### CODE REVIEW

**Comment by louistrue** [comment]
Thanks — all review feedback addressed in a94651d7, plus an adversarial pass that found one more real bug.

**CodeRabbit**
1. **Constructor overflow** (Major) — fixed with `checked_add`/`checked_mul` returning `Err("ringLengths overflow")`. This was a genuine wasm32 trap: `[3, 4294967293]` wrapped the u32 vertex total to 0, passed the length check, then panicked slicing `coords`. The 64-bit host summed correctly, which is why it stayed hidden locally.
2. **Sanitize constructor rings** (Minor) — the constructor and `fromMeshOutline` now sanitize, so `isEmpty`/`bounds()` report the same rings a boolean keeps. `new([], [0])` and friends now come back genuinely empty.
3. **Test title mismatch** (Minor) — renamed to `difference2d punches a hole and reports the holed shape as one region` (no `union2d` in the body).

**Codex** — documented `Contours2D` + `union2d`/`difference2d`/`intersection2d`/`resolve2d` in `docs/api/wasm.md` (winding contract, island preservation, manual-free accumulator loop), per the same-PR documentation-upkeep rule.

**Adversarial Fable pass** — escalated finding 2: a **3-vertex collinear** ring is structurally valid (≥3 finite vertices) but covers zero area, so it slipped past the vertex-count sanitation and left `isEmpty`/`bounds()` disagreeing with a later boolean. `sanitize` now also drops exactly-zero-area rings — behaviour-preserving, since i_overlay emits nothing for them anyway and a real thin sliver has non-zero area. Also tightened two doc overreaches it flagged: the "input winding respected, not normalised" claim (a lone CW ring fills under NonZero — clarified winding is relative to the regions a ring overlaps) and `coords()` no longer claims to be the "exact inverse" of the constructor now that sanitation strips closing/degenerate input.

New coverage: Rust asserts `sanitize` drops a collinear ring; the boundary suite (36 pass) asserts the constructor drops degenerate + collinear rings so the accessors stay consistent.

**Comment by louistrue** [comment]
@coderabbitai review

**Comment by louistrue** [comment]
Re-review addressed in 98290dd1.

**Major — zero-shoelace bow-tie (CodeRabbit, correct and sharp):** the previous commit's `is_zero_area` dropped rings by *signed* area, which is wrong — a self-intersecting bow-tie has zero signed area but both lobes fill under NonZero. I verified this directly against i_overlay: `[[0,0],[2,2],[2,0],[0,2]]` resolves to two shapes covering area 2 (an hourglass covers 6), so the area drop was silently discarding real coverage. Replaced with an **exact collinearity test** (all vertices on one line via a zero cross-product): it still drops a genuinely degenerate line ring — the accessor-consistency case — but keeps every bow-tie for i_overlay to fill, and being exact it never touches a real thin sliver. Added a regression test asserting the zero-signed-area bow-tie is kept and covers area 2.

**Winding doc (`wasm.md`):** clarified — under NonZero a CW ring subtracts only where it overlaps positive winding; a nested CW ring is a hole, a lone CW ring still fills.

**Perf verdict (nitpick):** the PR body carries one. There is no worker-pool measurement to add because this path is not on the worker-pool pipeline: `contour_bool2d` is a new leaf module reached only through the new `@ifc-lite/wasm` exports — `produce_element_meshes` / `processGeometryBatch` never call it. There is no end-to-end geometry-processing path to measure; the operations are `i_overlay` (the same engine the void/CSG paths already use), f64, deterministic. Skipping the worker-pool measurement as not-applicable.

**Two findings that re-posted but are already resolved:**
- *Sanitize accepted constructor rings* (`bool2d.rs`): the constructor and `fromMeshOutline` already call `sanitize_contours` since a94651d7. `new([], Uint32Array.of(0))` builds one empty ring which sanitation drops, so `isEmpty` is `true`. Boundary test `the constructor drops degenerate rings so accessors stay consistent` covers it.
- *Document the new WASM boolean API* (Codex): `docs/api/wasm.md` gained a full "2D Boolean Operations (contour sets)" section in a94651d7 — winding contract, island preservation, and the manual-free accumulator loop.

**Comment by louistrue** [comment]
@coderabbitai review

**Comment by louistrue** [comment]
eb925ea0: aligned the remaining doc comments (constructor, sanitize re-export, docs/api/wasm.md) to say the drop is `< 3 vertices / non-finite / exactly collinear`, matching the implementation now that a zero-area bow-tie is kept. Comment-only.

The other two inline re-posts are already resolved: the constructor sanitation landed in a94651d7 (CodeRabbit marked it addressed), and the `docs/api/wasm.md` boolean-API section landed in the same commit.

@coderabbitai review

**Comment by louistrue** [comment]
8b545801: CodeRabbit CLI recovered and flagged one thing the hosted review didn't (neither is a superset) — the winding clarification hadn't reached the changeset prose, which still said clockwise rings 'remove' area unconditionally. Fixed to match the module + API-guide wording (a lone CW ring fills; CW makes a hole only where it cancels positive winding).

@coderabbitai review

**Comment by louistrue** [comment]
@coderabbitai full review

File: packages/wasm/pkg/ifc-lite.d.ts [hunk 1]
```diff
@@ -35,6 +35,71 @@ export class ClashSession {
   runRule(group_a: Uint32Array, group_b: Uint32Array, mode: number, tolerance: number, clearance: number, report_touch: boolean): ClashRunResult;
 }
 
+export class Contours2D {
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/wasm/pkg/ifc-lite.d.ts:38]: `docs/api/wasm.md` does now document this API: `Contours2D` (including `fromMeshOutline`) is listed in the exported-handle paragraph and given its own section, with `union2d` / `difference2d` / `intersection2d` / `resolve2d` signatures, the NonZero winding contract, the keeps-every-disjoint-shape behaviour, and the manual `free()` requirement with a worked accumulating-loop example. Resolving as verified against the current head.

**Comment by mehmet-ylcnky** [comment]
@louistrue,
Thanks for the great support 🙏
I implemented the suggestions of Design 1 in our discussion which already produced great result.
I will now continue with Design 2 with the outputs of this PR.
Let's keep in touch 🤜🤛
